components/script/dom/paintworkletglobalscope.rs

```diff
@@ -54,44 +89,232 @@ impl PaintWorkletGlobalScope {
-            self.buffer.borrow_mut().truncate(new_buffer_size);
+                          size: Size2D<Au>,
+                          sender: IpcSender<CanvasData>)
+    {
+        // TODO: document paint definitions.
+        self.invoke_a_paint_callback(name, size, sender);
+    }
+
+    /// https://drafts.css-houdini.org/css-paint-api/#invoke-a-paint-callback
+    #[allow(unrooted_must_root)]
+    #[allow(unsafe_code)]
+    fn invoke_a_paint_callback(&self,
+                               name: Atom,
+                               size: Size2D<Au>,
+                               sender: IpcSender<CanvasData>)
+    {
+        let width = size.width.to_px().abs() as u32;
+        let height = size.height.to_px().abs() as u32;
+        debug!("Invoking a paint callback {}({},{}).", name, width, height);
+
+        let cx = self.worklet_global.get_cx();
+        let _ac = JSAutoCompartment::new(cx, self.worklet_global.reflector().get_jsobject().get());
+
+        // TODO: Steps 1-2.1.
+        // Step 3.
+        debug!("Looking up paint definition {}.", name);
+        let definition = match self.paint_definitions.borrow().get(&name) {
+            Some(defn) => defn.clone(),
+            None => {
+                // Step 2.2.
+                warn!("Drawing un-registered paint definition {}.", name);
+                return self.send_invalid_image(size, sender);
+            }
+        };
+
+        // Steps 4-5
+        // TODO: the spec requires calling the constructor now, but we might want to
+        // prepopulate the paint instance in `RegisterPaint`, to avoid calling it in
+        // the primary worklet thread.
+        rooted!(in(cx) let mut paint_instance = UndefinedValue());
+        match self.paint_class_instances.borrow_mut().entry(name.clone()) {
+            Entry::Occupied(entry) => paint_instance.set(entry.get().get()),
+            Entry::Vacant(entry) => {
+                debug!("Creating new paint instance.");
+                // Step 5.1
+                if !definition.constructor_valid_flag.get() {
+                    debug!("Drawing invalid paint definition {}.", name);
+                    return self.send_invalid_image(size, sender);
+                }
+                // Step 5.2-5.3
+                let args = HandleValueArray::new();
+                rooted!(in(cx) let mut result = null_mut());
+                unsafe { Construct1(cx, definition.class_constructor.handle(), &args, result.handle_mut()); }
+                paint_instance.set(ObjectValue(result.get()));
+                if unsafe { JS_IsExceptionPending(cx) } {
+                    debug!("Paint constructor threw an exception {}.", name);
+                    unsafe { JS_ClearPendingException(cx); }
+                    definition.constructor_valid_flag.set(false);
+                    return self.send_invalid_image(size, sender);
+                }
+                // Step 5.4
+                entry.insert(Heap::new(paint_instance.get()));
```

Comment:
I'm just going to flag this now - we must not introduce any new uses of Heap::new into Servo due to https://github.com/servo/rust-mozjs/issues/343 . Use Heap::default to create a Heap value and ensure it will never move after that point before calling Heap::set. This hashtable looks very scary to me for this reason :(